### PR TITLE
implicit-casts:false in flutter/lib/src/cupertino

### PR DESCRIPTION
**Pull request by a14n**
## Description

Enable `implicit-casts:false` in `flutter/lib/src/cupertino`

## Related Issues

None

## Tests

None because it's a refactoring.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 1]
```diff
@@ -271,7 +271,7 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
     Color backgroundColor,
     Color activeColor,
     Color inactiveColor,
-    Size iconSize,
+    double iconSize,
```
Comment (dnfield) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: I am confused about how this ever worked.

I am also hoping this doesn't end up being a breaking change under our new breaking change policy (/cc @Hixie).

File: packages/flutter/lib/src/cupertino/context_menu.dart [hunk 2]
```diff
@@ -990,7 +990,7 @@ class _ContextMenuRouteStaticState extends State<_ContextMenuRouteStatic> with T
       _moveAnimation = Tween<Offset>(
         begin: Offset.zero,
         end: Offset(
-          endX.clamp(-_kPadding, _kPadding),
+          endX.clamp(-_kPadding, _kPadding) as double,
```
Comment (dnfield) [inline packages/flutter/lib/src/cupertino/context_menu.dart]: ditto on question from other patch - I think `.toDouble` would be safer here.

File: packages/flutter/lib/src/cupertino/route.dart [hunk 3]
```diff
@@ -721,16 +721,16 @@ class _CupertinoEdgeShadowDecoration extends Decoration {
 
   @override
   _CupertinoEdgeShadowDecoration lerpFrom(Decoration a, double t) {
-    if (a is! _CupertinoEdgeShadowDecoration)
-      return _CupertinoEdgeShadowDecoration.lerp(null, this, t);
-    return _CupertinoEdgeShadowDecoration.lerp(a, this, t);
+    if (a is _CupertinoEdgeShadowDecoration)
+      return _CupertinoEdgeShadowDecoration.lerp(a, this, t);
+    return _CupertinoEdgeShadowDecoration.lerp(null, this, t);
```
Comment (dnfield) [inline packages/flutter/lib/src/cupertino/route.dart]: Do these have to be changed?  If not, should we move this to a separate patch?

File: packages/flutter/lib/src/cupertino/route.dart [hunk 4]
```diff
@@ -721,16 +721,16 @@ class _CupertinoEdgeShadowDecoration extends Decoration {
...
   }
 
   @override
   _CupertinoEdgeShadowDecoration lerpTo(Decoration b, double t) {
-    if (b is! _CupertinoEdgeShadowDecoration)
-      return _CupertinoEdgeShadowDecoration.lerp(this, null, t);
-    return _CupertinoEdgeShadowDecoration.lerp(this, b, t);
+    if (b is _CupertinoEdgeShadowDecoration)
+      return _CupertinoEdgeShadowDecoration.lerp(this, b, t);
+    return _CupertinoEdgeShadowDecoration.lerp(this, null, t);
```
Comment (dnfield) [inline packages/flutter/lib/src/cupertino/route.dart]: Ditto

File: packages/flutter/lib/src/cupertino/sliding_segmented_control.dart [hunk 5]
```diff
@@ -686,9 +686,9 @@ class _RenderSegmentedControl<T> extends RenderBox
     return childCount == 0
       ? null
       // This assumes all children have the same width.
-      : (location.dx / (size.width / childCount))
+      : ((location.dx / (size.width / childCount))
         .floor()
-        .clamp(0, childCount - 1);
+        .clamp(0, childCount - 1) as int);
```
Comment (dnfield) [inline packages/flutter/lib/src/cupertino/sliding_segmented_control.dart]: `.toInt()`

File: packages/flutter/lib/src/cupertino/segmented_control.dart [hunk 6]
```diff
@@ -541,7 +541,8 @@ class _RenderSegmentedControl<T> extends RenderBox
     RenderBox child = firstChild;
     double minWidth = 0.0;
     while (child != null) {
-      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      final _SegmentedControlContainerBoxParentData childParentData =
+        child.parentData as _SegmentedControlContainerBoxParentData;
```
Comment (dnfield) [inline packages/flutter/lib/src/cupertino/segmented_control.dart]: Here and elsewhere: these should still be one line from my understanding of the Flutter style guide.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart (hunk 1 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/bottom_tab_bar.dart]: To be a breaking change there should be some code that call this method with a `Size`. But call it with a `Size` would always fail. So I'll arg this is not a breaking change :)

File: packages/flutter/lib/src/cupertino/route.dart (hunk 3 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/route.dart]: By changing the order `a` can be promoted to `_CupertinoEdgeShadowDecoration` and avoid an explicit cast if the order remain the same.

Note that with changes regarding type promotion in NNBD the previous code would have working without changes.

File: packages/flutter/lib/src/cupertino/segmented_control.dart (hunk 6 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/segmented_control.dart]: no problem to put it on one line. I usually split when the line length is more that 120 cols.

Comment (dnfield) [reply packages/flutter/lib/src/cupertino/segmented_control.dart]: That makes sense, but there's a preference to avoid linebreaks for assignments (https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters)

**Review by dnfield — APPROVED** [review]
LGTM modulo putting the assignments on one line.

As discussed elsewhere, the `toInt` and `toDouble` comments should be resolved by dart-lang/sdk#39652

File: packages/flutter/lib/src/cupertino/segmented_control.dart (hunk 6 above)
Comment (a14n) [reply packages/flutter/lib/src/cupertino/segmented_control.dart]: Done
